src/utils.c

```diff
@@ -67,3 +71,98 @@ void rv_clock_gettime(struct timespec *tp)
     tp->tv_sec = tv_sec;
     tp->tv_nsec = tv_usec / 1000; /* Transfer to microseconds */
 }
+
+char *sanitize_path(const char *input)
+{
+    size_t n = strnlen(input, MAX_PATH_LEN);
+
+    char *ret = malloc(MAX_PATH_LEN + 1);
+    memset(ret, '\0', MAX_PATH_LEN + 1);
+
+    /* After sanitization, the new path will only be shorter than the original
+     * one. Thus, we can reuse the space */
+    if (n == 0) {
+        ret[0] = '.';
+        return ret;
+    }
+
+    int rooted = (input[0] == '/');
+
+    /*
+     * Invariants:
+     *	reading from path; r is index of next byte to process -> path[r]
+     *	writing to buf; w is index of next byte to write -> ret[strlen(ret)]
+     *	dotdot is index in buf where .. must stop, either because
+     *		a) it is the leading slash
+     *      b) it is a leading ../../.. prefix.
+     */
+    size_t w = 0;
+    size_t r = 0;
+    size_t dotdot = 0;
+    if (rooted) {
+        ret[w] = '/';
+        w++;
+        r = 1;
+        dotdot = 1;
+    }
+
+    while (r < n) {
+        if (input[r] == '/') {
+            /*  empty path element */
+            r++;
+        } else if (input[r] == '.' &&
+                   (r + 1 == n || input[r + 1] == '/')) {
+            /* . element */
+            r++;
+        } else if (input[r] == '.' && input[r + 1] == '.' &&
+                   (r + 2 == n || input[r + 2] == '/')) {
+            /* .. element: remove to last / */
+            r += 2;
+
+            if (w > dotdot) {
+                /* can backtrack */
+                w--;
+                while (w > dotdot && ret[w] != '/') {
+                    w--;
+                }
+            } else if (!rooted) {
+                /* cannot backtrack, but not rooted, so append .. element. */
+                if (w > 0) {
+                    ret[w] = '/';
+                    w++;
+                }
+                ret[w] = '.';
+                w++;
+                ret[w] = '.';
+                w++;
+                dotdot = w;
+            }
+        } else {
+            /* real path element.
+               add slash if needed */
+            if ((rooted && w != 1) || (!rooted && w != 0)) {
+                ret[w] = '/';
+                w++;
+            }
+
+            /* copy element */
+            for (; r < n && input[r] != '/'; r++) {
+                ret[w] = input[r];
+                w++;
+            }
+        }
+        // printf("w = %ld, r = %ld, dotdot = %ld\nret = %s\n", w, r, dotdot,
+        // ret);
+    }
+
+    /* Turn empty string into "." */
+    if (w == 0) {
+        ret[w] = '.';
+        w++;
+    }
+
+    for (size_t i = w; i < n; i++) {
+        ret[i] = '\0';
```

Comment:
Why not using `memset`, which is usually optimized by libc?

Reply:
Not properly thought of :( Implemented with `memset` now